File: frontend/lib/features/workspace/models/workspace.dart
// features/workspace/models/workspace.dart
// Purpose: Defines the Workspace model and workspace status enum.
// Responsibilities: Handles JSON serialization and status mapping.

enum WorkspaceStatus {
  ready,
  processing,
  failed;

  static WorkspaceStatus fromString(String val) {
    switch (val.toLowerCase()) {
      case 'processing':
        return WorkspaceStatus.processing;
      case 'failed':
        return WorkspaceStatus.failed;
      case 'ready':
      default:
        return WorkspaceStatus.ready;
    }
  }

  String toJson() => name;
}

class Workspace {
  final String id;
  final String name;
  final DateTime createdAt;
  final WorkspaceStatus status;
  final int sourcesCount;
  final String instructions;
  final int sizeBytes;
  final String? errorMessage;

  Workspace({
    required this.id,
    required this.name,
    required this.createdAt,
    required this.status,
    required this.sourcesCount,
    this.instructions = '',
    this.sizeBytes = 0,
    this.errorMessage,
  });

  factory Workspace.fromJson(Map<String, dynamic> json) {
    return Workspace(
      id: json['id'] as String,
      name: json['name'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      status: WorkspaceStatus.fromString(json['status'] as String? ?? 'ready'),
      sourcesCount: json['sources_count'] as int? ?? 0,
      instructions: json['instructions'] as String? ?? '',
      sizeBytes: json['size_bytes'] as int? ?? 0,
      errorMessage: json['error_message'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'created_at': createdAt.toIso8601String(),
      'status': status.toJson(),
      'sources_count': sourcesCount,
      'instructions': instructions,
      'size_bytes': sizeBytes,
      'error_message': errorMessage,
    };
  }

  Workspace copyWith({
    String? id,
    String? name,
    DateTime? createdAt,
    WorkspaceStatus? status,
    int? sourcesCount,
    String? instructions,
    int? sizeBytes,
    String? errorMessage,
  }) {
    return Workspace(
      id: id ?? this.id,
      name: name ?? this.name,
      createdAt: createdAt ?? this.createdAt,
      status: status ?? this.status,
      sourcesCount: sourcesCount ?? this.sourcesCount,
      instructions: instructions ?? this.instructions,
      sizeBytes: sizeBytes ?? this.sizeBytes,
      errorMessage: errorMessage ?? this.errorMessage,
    );
  }
}

File: backend/app/models/workspace.py
# app/models/workspace.py
# Purpose: Pydantic schemas for Workspace CRUD operations.
# Responsibilities: Defines validation models for HTTP requests/responses.

from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional

class WorkspaceBase(BaseModel):
    name: str = Field(..., description="The name of the workspace", min_length=1, max_length=100)
    instructions: Optional[str] = Field("", description="Custom workspace system instructions")

class WorkspaceCreate(WorkspaceBase):
    pass

class WorkspaceRename(WorkspaceBase):
    pass

class WorkspaceUpdate(BaseModel):
    name: Optional[str] = Field(None, description="The name of the workspace", min_length=1, max_length=100)
    instructions: Optional[str] = Field(None, description="Custom workspace system instructions")

class Workspace(WorkspaceBase):
    id: str = Field(..., pattern=r"^[0-9a-f-]{36}$", description="Unique workspace identifier")
    created_at: datetime = Field(..., description="Timestamp when workspace was created")
    status: str = Field("ready", description="Status of workspace processing (ready, processing, failed)")
    sources_count: int = Field(0, description="Total number of sources added to the workspace")

    class Config:
        json_schema_extra = {
            "example": {
                "id": "e4f8d223-9f8e-4a6c-905c-e6593539a2f1",
                "name": "My Research Workspace",
                "created_at": "2026-06-16T22:00:00.000Z",
                "status": "ready",
                "sources_count": 2
            }
        }
